File: Data Structures & Algorithms/add-two-numbers/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def addTwoNumbers(self, l1, l2):

        dummy = ListNode(0)
        curr = dummy

        carry = 0

        while l1 or l2 or carry:

            # Get digits, using 0 if a list is finished
            digit1 = l1.val if l1 else 0
            digit2 = l2.val if l2 else 0

            # Add the digits and carry
            total = digit1 + digit2 + carry

            # Current digit
            digit = total % 10

            # Carry for next position
            carry = total // 10

            # Create the new node
            curr.next = ListNode(digit)
            curr = curr.next

            # Move through the input lists
            if l1:
                l1 = l1.next

            if l2:
                l2 = l2.next

        return dummy.next

File: Data Structures & Algorithms/reverse-nodes-in-k-group/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {

    public ListNode reverseKGroup(ListNode head, int k) {

        ListNode dummy = new ListNode(0);
        dummy.next = head;

        ListNode groupPrev = dummy;

        while (true) {

            // Find the kth node
            ListNode kth = getKth(groupPrev, k);

            // Not enough nodes left
            if (kth == null) {
                break;
            }

            // Node after the current group
            ListNode groupNext = kth.next;

            // Reverse the group
            ListNode prev = groupNext;
            ListNode curr = groupPrev.next;

            while (curr != groupNext) {

                ListNode temp = curr.next;
                curr.next = prev;
                prev = curr;
                curr = temp;
            }

            // Connect previous group to reversed group
            ListNode temp = groupPrev.next;
            groupPrev.next = kth;

            // Move groupPrev to the end of the reversed group
            groupPrev = temp;
        }

        return dummy.next;
    }

    private ListNode getKth(ListNode curr, int k) {

        while (curr != null && k > 0) {
            curr = curr.next;
            k--;
        }

        return curr;
    }
}